### ISSUE TITLE
Streaming decoder

### ISSUE THREAD

**Issue by beatgammit** [go-yaml/yaml#4]
It would be nice to have a `Decoder.Decode` like `encoding/json` for processing streams. I basically want to do:

```
f, _ := os.Open(filename)
yaml.NewDecoder(f).Decode(&obj)
```

Instead of:

```
b, _ := ioutil.ReadFile(filename)
yaml.Unmarshal(b, &obj)
```

This makes it nicer for decoding `yaml` documents from `io.Reader`s without having to read everything into a `[]byte` first.

**Comment by niemeyer** [comment]
This should also consider streams with multiple documents.

**Comment by james-lawrence** [comment]
+1

**Comment by at15** [comment]
There is a PR https://github.com/go-yaml/yaml/pull/163/ for this issue, but got stuck on CLA

**Comment by im-kulikov** [comment]
#288 Workaround Encoder/Decoder

**Comment by rogpeppe** [comment]
Fixed by #301
